File: Sources/ManageArmsCore/ProcessScanner.swift
import Foundation

/// `ps` の 1 行。
public struct ProcessRow: Equatable, Sendable {
    public let pid: Int32
    public let ppid: Int32
    /// `etime` そのまま。`18:54` / `02:10:33` / `3-04:05:06`。
    /// 自前で秒に直さない — 表示にしか使わないうえ、書式が増えても壊れない。
    public let elapsed: String
    public let command: String

    public init(pid: Int32, ppid: Int32, elapsed: String, command: String) {
        self.pid = pid; self.ppid = ppid; self.elapsed = elapsed; self.command = command
    }
}

/// 実行中の MCP サーバー 1 件。
public struct RunningMCP: Equatable, Sendable {
    /// 親を辿って特定できたエージェント。辿り着けなければ `nil`。
    public let owner: Agent?
    public let elapsed: String
    public let pid: Int32
}

/// MCP が実際に起動しているかを見る。DESIGN.md 3.9。
///
/// **MCP だけが「今この瞬間の状態」を持つ。** Skills / Plugins には実行実体が無く、
/// 「使用中」は 1 ターンで消えるイベントにしかならない（そちらは `UsageScanner`）。
///
/// `ps` を 1 回叩くだけなので常駐もポーリングも要らない。
/// `didBecomeActive` の再スキャンに相乗りさせる（3.5 と衝突しない）。
public enum ProcessScanner {

    public static func snapshot(env: Environment) -> [ProcessRow] {
        guard let out = try? env.run(["ps", "-eo", "pid,ppid,etime,command"]) else { return [] }
        return parse(out)
    }

    /// `  87139 87070       18:58 Cursor Helper: mcp-process`
    /// コマンドは空白を含むので、先頭 3 列だけ切り出して残りを丸ごと使う。
    static func parse(_ output: String) -> [ProcessRow] {
        output.split(separator: "\n").compactMap { line in
            let fields = line.split(maxSplits: 3, omittingEmptySubsequences: true) {
                $0 == " " || $0 == "\t"
            }
            guard fields.count == 4,
                  let pid = Int32(fields[0]), let ppid = Int32(fields[1])
            else { return nil }                       // ヘッダ行はここで落ちる
            return ProcessRow(pid: pid, ppid: ppid,
                              elapsed: String(fields[2]), command: String(fields[3]))
        }
    }

    // MARK: - 登録済みサーバーとの突き合わせ

    /// 「MCP らしいプロセス」を当てにいくのではなく、**登録済みの設定から探す**。
    /// UI が答えたい問いは「この登録済みサーバーは動いているか」であって
    /// 「動いている MCP を全部挙げよ」ではない。
    public static func running(_ servers: [MCPServer], in rows: [ProcessRow])
        -> [String: RunningMCP]
    {
        var result: [String: RunningMCP] = [:]
        for server in servers {
            let tokens = signatures(of: server)
            guard !tokens.isEmpty else { continue }
            let matched = rows.filter { row in tokens.contains { row.command.contains($0) } }
            guard !matched.isEmpty else { continue }

            // 同じサーバーが npm exec → 本体 → watchdog と数珠つなぎになる（実測）。
            // 一致した集合の中で親を持たないものが起点で、稼働時間もそれが正しい。
            let pids = Set(matched.map(\.pid))
            let root = matched.first { !pids.contains($0.ppid) } ?? matched[0]
            result[server.name] = RunningMCP(
                owner: owner(of: root, in: rows), elapsed: root.elapsed, pid: root.pid)
        }
        return result
    }

    /// プロセスを一意に指す語。`npx -y chrome-devtools-mcp@latest` なら
    /// `chrome-devtools-mcp` が拾える。
    ///
    /// **`npx` や `-y` のような共通語で照合してはいけない** — 無関係なプロセスに当たる。
    /// 3 文字以下も切る（`.` や `uv` が全部に一致するため）。
    static func signatures(of server: MCPServer) -> [String] {
        var tokens: [String] = []
        for token in server.args where !token.hasPrefix("-") {
            tokens.append(stripVersion(token))
        }
        if let command = server.command {
            tokens.append(stripVersion((command as NSString).lastPathComponent))
        }
        if case .http(let url, _) = server.transport { tokens.append(url) }
        // サーバー名そのものは最後の手掛かり。設定に固有の語が無いことがある。
        tokens.append(server.name)
        return tokens.filter { $0.count > 3 && !generic.contains($0) }
    }

    /// パッケージ実行系はコマンド行に現れても対象を絞らない。
    static let generic: Set<String> = [
        "npx", "npm", "exec", "node", "uvx", "uv", "run", "python", "python3",
        "deno", "bunx", "docker", "pipx", "-y", "--yes", "latest",
    ]

    /// `chrome-devtools-mcp@latest` → `chrome-devtools-mcp`。
    /// スコープ付き（`@scope/pkg@1.2.3`）の先頭 `@` は落とさない。
    static func stripVersion(_ token: String) -> String {
        guard let at = token.lastIndex(of: "@"), at != token.startIndex else { return token }
        return String(token[token.startIndex..<at])
    }

    // MARK: - 所属エージェントの特定

    /// 親を辿ってエージェントに行き着くか見る。実測:
    /// `chrome-devtools-mcp` → `npm exec` → `Cursor Helper: mcp-process` → `Cursor.app`
    static func owner(of row: ProcessRow, in rows: [ProcessRow]) -> Agent? {
        let byPID = Dictionary(rows.map { ($0.pid, $0) }, uniquingKeysWith: { a, _ in a })
        var current: ProcessRow? = row
        var depth = 0
        while let node = current, depth < 24 {         // 循環しても抜けられるようにする
            if let agent = agent(ofCommand: node.command) { return agent }
            guard node.ppid > 1 else { return nil }
            current = byPID[node.ppid]
            depth += 1
        }
        return nil
    }

    /// **実行ファイル名を先に見る。** Codex は Cursor の拡張の中に入っていることがあり
    /// （実測: `~/.cursor/extensions/openai.chatgpt-…/bin/…/codex`）、
    /// パスに `cursor` が含まれるからと先に Cursor と判定すると取り違える。
    ///
    /// 手掛かりは `Agent.cliName` / `Agent.processMarkers` が持つ。
    /// ここでエージェント名を直書きすると、増えたぶんが黙って `nil` になる。
    static func agent(ofCommand command: String) -> Agent? {
        let executable = command.split(separator: " ").first.map(String.init) ?? command
        let name = (executable as NSString).lastPathComponent
        if let byCLI = Agent.allCases.first(where: { $0.cliName == name }) { return byCLI }
        return Agent.allCases.first { agent in
            agent.processMarkers.contains { command.contains($0) }
        }
    }
}

File: src/processScanner.ts
import { basename } from "node:path";
import { AgentId, AGENT_IDS, cliName, processMarkers } from "./agent";
import { Run } from "./env";
import { MCPServer } from "./mcpServer";

/** プロセス一覧の 1 行。`elapsed` は表示にしか使わないので秒へ直さない。 */
export type ProcessRow = {
  readonly pid: number;
  readonly ppid: number;
  readonly elapsed: string;
  readonly command: string;
};

export type RunningMCP = {
  /** 親を辿って特定できたエージェント。辿り着けなければ null。 */
  readonly owner: AgentId | null;
  readonly elapsed: string;
  readonly pid: number;
};

/**
 * 全プロセスを 1 回だけ取る。OS ごとにコマンドは違うが、
 * 「pid / ppid / 経過 / コマンド行」の 4 列に揃えてから先は同じ処理にする。
 */
export async function snapshot(run: Run): Promise<ProcessRow[]> {
  if (process.platform === "win32") {
    // tasklist はコマンド行を持たないので、MCP の判別に使えない。
    const script = "Get-CimInstance Win32_Process | Select-Object ProcessId,ParentProcessId,"
      + "CreationDate,CommandLine | ConvertTo-Json -Compress";
    const output = await run(["powershell", "-NoProfile", "-Command", script]).catch(() => "");
    return parseWindows(output);
  }
  const output = await run(["ps", "-eo", "pid,ppid,etime,command"]).catch(() => "");
  return parsePs(output);
}

/** `  87139 87070       18:58 Cursor Helper: mcp-process` */
export function parsePs(output: string): ProcessRow[] {
  return output.split("\n").flatMap(line => {
    // コマンドは空白を含むので、先頭 3 列だけ切り出して残りを丸ごと使う。
    const match = /^\s*(\d+)\s+(\d+)\s+(\S+)\s+(.*)$/.exec(line);
    if (match === null) return [];      // ヘッダ行はここで落ちる
    return [{ pid: Number(match[1]), ppid: Number(match[2]), elapsed: match[3], command: match[4] }];
  });
}

export function parseWindows(output: string): ProcessRow[] {
  let parsed: unknown;
  try {
    parsed = JSON.parse(output);
  } catch {
    return [];
  }
  const list = Array.isArray(parsed) ? parsed : [parsed];
  return list.flatMap(raw => {
    if (typeof raw !== "object" || raw === null) return [];
    const item = raw as Record<string, unknown>;
    if (typeof item.ProcessId !== "number" || typeof item.CommandLine !== "string") return [];
    return [{
      pid: item.ProcessId,
      ppid: typeof item.ParentProcessId === "number" ? item.ParentProcessId : 0,
      elapsed: typeof item.CreationDate === "string" ? item.CreationDate : "",
      command: item.CommandLine,
    }];
  });
}

/** パッケージ実行系はコマンド行に現れても対象を絞らない。 */
const GENERIC: ReadonlySet<string> = new Set([
  "npx", "npm", "exec", "node", "uvx", "uv", "run", "python", "python3",
  "deno", "bunx", "docker", "pipx", "-y", "--yes", "latest",
]);

/** `chrome-devtools-mcp@latest` → `chrome-devtools-mcp`。スコープ付きの先頭 `@` は残す。 */
export function stripVersion(token: string): string {
  const at = token.lastIndexOf("@");
  return at > 0 ? token.slice(0, at) : token;
}

/**
 * プロセスを一意に指す語。`npx -y chrome-devtools-mcp@latest` なら
 * `chrome-devtools-mcp` が拾える。`npx` や `-y` のような共通語では照合しない —
 * 無関係なプロセスに当たる。3 文字以下も切る（`uv` が全部に一致する）。
 */
export function signatures(server: MCPServer): string[] {
  const tokens: string[] = [];
  if (server.transport.type === "stdio") {
    for (const token of server.transport.args) {
      if (!token.startsWith("-")) tokens.push(stripVersion(token));
    }
    tokens.push(stripVersion(basename(server.transport.command)));
  } else {
    tokens.push(server.transport.url);
  }
  // サーバー名そのものは最後の手掛かり。設定に固有の語が無いことがある。
  tokens.push(server.name);
  return tokens.filter(token => token.length > 3 && !GENERIC.has(token));
}

/**
 * 登録済みの設定から探す。UI が答えたい問いは「この登録済みサーバーは動いているか」で、
 * 「動いている MCP を全部挙げよ」ではない。
 */
export function running(servers: MCPServer[], rows: ProcessRow[]): Map<string, RunningMCP> {
  const result = new Map<string, RunningMCP>();
  for (const server of servers) {
    if (!server.enabled) continue;
    const tokens = signatures(server);
    if (tokens.length === 0) continue;
    const matched = rows.filter(row => tokens.some(token => row.command.includes(token)));
    if (matched.length === 0) continue;

    // 同じサーバーが npm exec → 本体 → watchdog と数珠つなぎになる（実測）。
    // 一致した集合の中で親を持たないものが起点で、稼働時間もそれが正しい。
    const pids = new Set(matched.map(row => row.pid));
    const root = matched.find(row => !pids.has(row.ppid)) ?? matched[0];
    result.set(server.name, { owner: owner(root, rows), elapsed: root.elapsed, pid: root.pid });
  }
  return result;
}

/**
 * 親を辿ってエージェントに行き着くか見る。実測:
 * `chrome-devtools-mcp` → `npm exec` → `Cursor Helper: mcp-process` → `Cursor.app`
 */
export function owner(row: ProcessRow, rows: ProcessRow[]): AgentId | null {
  const byPid = new Map(rows.map(item => [item.pid, item]));
  let current: ProcessRow | undefined = row;
  for (let depth = 0; current !== undefined && depth < 24; depth++) {  // 循環しても抜ける
    const agent = agentOfCommand(current.command);
    if (agent !== null) return agent;
    if (current.ppid <= 1) return null;
    current = byPid.get(current.ppid);
  }
  return null;
}

/**
 * 実行ファイル名を先に見る。Codex は Cursor の拡張の中に入っていることがあり、
 * パスに `cursor` が含まれるからと先に Cursor と判定すると取り違える。
 */
export function agentOfCommand(command: string): AgentId | null {
  const executable = command.split(" ")[0] ?? command;
  const name = basename(executable).replace(/\.(exe|cmd|bat)$/i, "");
  const byCli = AGENT_IDS.find(agent => cliName(agent) === name);
  if (byCli !== undefined) return byCli;
  return AGENT_IDS.find(agent => processMarkers(agent).some(marker => command.includes(marker))) ?? null;
}

/** View 表示中の 3 秒ポーリングから呼ぶ。key は "agent:serverName"。 */
export async function mcpStatus(servers: Partial<Record<AgentId, MCPServer[]>>, run: Run):
  Promise<Record<string, boolean>> {
  const rows = await snapshot(run);
  const status: Record<string, boolean> = {};
  for (const agent of AGENT_IDS) {
    const live = running(servers[agent] ?? [], rows);
    for (const server of servers[agent] ?? []) {
      status[`${agent}:${server.name}`] = live.has(server.name);
    }
  }
  return status;
}
